apps/src/maze/mazeController.js

```diff
@@ -0,0 +1,363 @@
+    timeoutList.clearTimeouts();
+
+    // Move Pegman into position.
+    this.pegmanX = this.subtype.start.x;
+    this.pegmanY = this.subtype.start.y;
+
+    this.pegmanD = this.startDirection;
+    this.animationsController.reset(first);
+
+    // Move the init dirt marker icons into position.
+    this.map.resetDirt();
+    this.resetDirtImages(false);
+
+    // Reset the obstacle image.
+    var obsId = 0;
+    var x, y;
+    for (y = 0; y < this.map.ROWS; y++) {
+      for (x = 0; x < this.map.COLS; x++) {
+        var obsIcon = document.getElementById('obstacle' + obsId);
+        if (obsIcon) {
+          obsIcon.setAttributeNS('http://www.w3.org/1999/xlink', 'xlink:href',
+                                 this.skin.obstacleIdle);
+        }
+        ++obsId;
+      }
+    }
+
+    if (this.subtype.resetTiles) {
+      this.subtype.resetTiles();
+    } else {
+      this.resetTiles_();
+    }
+  }
+
+  resetTiles_() {
+    // Reset the tiles
+    var tileId = 0;
+    for (var y = 0; y < this.map.ROWS; y++) {
+      for (var x = 0; x < this.map.COLS; x++) {
+        // Tile's clipPath element.
+        var tileClip = document.getElementById('tileClipPath' + tileId);
+        tileClip.setAttribute('visibility', 'visible');
+        // Tile sprite.
+        var tileElement = document.getElementById('tileElement' + tileId);
+        tileElement.setAttributeNS(
+            'http://www.w3.org/1999/xlink', 'xlink:href', this.skin.tiles);
+        tileElement.setAttribute('opacity', 1);
+        tileId++;
+      }
+    }
+  }
+
+  animatedFinish(timePerStep) {
+    this.animationsController.scheduleDance(true, timePerStep);
+  }
+
+  animatedMove(direction, timeForMove) {
+    var positionChange = tiles.directionToDxDy(direction);
+    var newX = this.pegmanX + positionChange.dx;
+    var newY = this.pegmanY + positionChange.dy;
+    this.animationsController.scheduleMove(newX, newY, timeForMove);
+    this.playAudio('walk');
+    this.pegmanX = newX;
+    this.pegmanY = newY;
+  }
+
+  animatedTurn(direction) {
+    var newDirection = this.pegmanD + direction;
+    this.animationsController.scheduleTurn(newDirection);
+    this.pegmanD = tiles.constrainDirection4(newDirection);
+  }
+
+  animatedFail(forward) {
+    var dxDy = tiles.directionToDxDy(this.pegmanD);
+    var deltaX = dxDy.dx;
+    var deltaY = dxDy.dy;
+
+    if (!forward) {
+      deltaX = -deltaX;
+      deltaY = -deltaY;
+    }
+
+    var targetX = this.pegmanX + deltaX;
+    var targetY = this.pegmanY + deltaY;
+    var frame = tiles.directionToFrame(this.pegmanD);
+    this.animationsController.displayPegman(
+      this.pegmanX + deltaX / 4,
+      this.pegmanY + deltaY / 4,
+      frame,
+    );
+    // Play sound and animation for hitting wall or obstacle
+    var squareType = this.map.getTile(targetY, targetX);
+    if (squareType === tiles.SquareType.WALL || squareType === undefined ||
+      (this.subtype.isScrat() && squareType === tiles.SquareType.OBSTACLE)) {
```

Comment:
😕 might be good to call into something like `this.subtype.handleHitObstacle` make the subtype override the behavior if it's different from the default.  (Future PR...)

Reply:
A thousand times yes. I would like in general to remove [these subtype identification methods](https://github.com/code-dot-org/code-dot-org/blob/14c057429f57ce467287ba44d8db18ad7be17ff0/apps/src/maze/subtype.js#L168-L186) entirely and update everything that's currently using them to instead leverage overridable methods